src/transformStateWithClones.js

```diff
@@ -7,7 +7,43 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const statesArray = [];
+
+  let currentState = { ...state };
+
+  actions.forEach((action) => {
+    switch (action.type) {
+      case 'clear':
+        currentState = {};
+        break;
+
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        break;
+
+      case 'removeProperties':
+        const keys = Object.keys(currentState);
+        const keysToRemove = keys.filter((key) => action.keysToRemove.includes(key));
+
+        const newState = {};
+
+        keys.forEach((key) => {
+          if (!keysToRemove.includes(key)) {
+            newState[key] = currentState[key];
+          }
+        });
+
+        currentState = { ...newState }
+        break;
+
+      default:
+        break;
+    }
+
+    statesArray.push({ ...currentState });
```

Comment:
You are correctly creating a new object for each state to avoid mutations on the original state object. However, it would be more efficient to push the `currentState` directly into `statesArray` after each action, as it is already a new object and doesn't need to be spread into another new object.